pkg/oauth/api/types.go

```diff
@@ -73,6 +73,34 @@ type OAuthClient struct {
 
 	// RedirectURIs is the valid redirection URIs associated with a client
 	RedirectURIs []string
+
+	// ScopeRestrictions describes which scopes this client can request.  Each requested scope
+	// is checked against each restriction.  If any restriction matches, then the scope is allowed.
+	// If no restriction matches, then the scope is denied.
+	ScopeRestrictions []ScopeRestriction
+
+	// AllowAnyScope indicates that the client is allowed to request a token unconstrained by scopes.
+	// If this is true, then ScopeRestrictions is ignored.
+	AllowAnyScope bool
```

Comment:
why do we need this? why wouldn't an empty `ScopeRestrictions` mean any scope is allowed? don't we need to do that for backwards compatibility anyway?